### PR TITLE
Lidar Driver Wrapper 

### PR DESCRIPTION
**Pull request by nickm1824**

### CODE REVIEW

**Comment by jalesseVolpe** [comment]
Great start. Have you attempted a colcon build? There are cpp dependencies and would like to know if they get resolved properly

**Comment by jalesseVolpe** [comment]
Andrew should review this

**Comment by asymingt** [comment]
> Great start. Have you attempted a colcon build? There are cpp dependencies and would like to know if they get resolved properly

I did a quick compile on the EC2 instance to see if there are any errors. Here's a log, so you can repeat the steps on your side. Note that I mount the package into the running instance, you one could vscode remote over SSH into EC2 and edit-test-edit-test without restarting the container.

```sh
asymingt@carma-ec2-cpu:~/development/carma-workspace$ cd c1tenth-driver-wrappers/
asymingt@carma-ec2-cpu:~/development/carma-workspace/c1tenth-driver-wrappers$ git pull
remote: Enumerating objects: 123, done.
remote: Counting objects: 100% (123/123), done.
remote: Compressing objects: 100% (113/113), done.
remote: Total 121 (delta 52), reused 4 (delta 1), pack-reused 0
Receiving objects: 100% (121/121), 29.16 KiB | 7.29 MiB/s, done.
Resolving deltas: 100% (52/52), done.
From github.com:KBR-CARMA/c1tenth-driver-wrappers
   529922d..26e22b9  c1tenth-develop      -> origin/c1tenth-develop
 * [new branch]      imu_driver_wrapper   -> origin/imu_driver_wrapper
 * [new branch]      lidar_driver_wrapper -> origin/lidar_driver_wrapper
Updating 529922d..26e22b9
Fast-forward
 README.md                       | 116 +-------------------------------------------------------------------------------------------
 docker/build-image.sh           | 153 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 docker/checkout.bash            |  62 ++++++++++++++++++++++++++++++++++++++++++++++++++
 docker/get-component-version.sh |  20 ++++++++++++++++
 docker/get-image-name.sh        |  24 +++++++++++++++++++
 docker/get-repo-name.sh         |  17 ++++++++++++++
 docker/get-system-version.sh    |  19 ++++++++++++++++
 docker/install.sh               |  30 ++++++++++++++++++++++++
 docker/system_release.sh        |  69 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 9 files changed, 395 insertions(+), 115 deletions(-)
 create mode 100644 docker/build-image.sh
 create mode 100644 docker/checkout.bash
 create mode 100644 docker/get-component-version.sh
 create mode 100644 docker/get-image-name.sh
 create mode 100644 docker/get-repo-name.sh
 create mode 100644 docker/get-system-version.sh
 create mode 100644 docker/install.sh
 create mode 100644 docker/system_release.sh
asymingt@carma-ec2-cpu:~/development/carma-workspace/c1tenth-driver-wrappers$ git checkout lidar_driver_wrapper 
Branch 'lidar_driver_wrapper' set up to track remote branch 'lidar_driver_wrapper' from 'origin'.
Switched to a new branch 'lidar_driver_wrapper'
asymingt@carma-ec2-cpu:~/development/carma-workspace$ cd ..
asymingt@carma-ec2-cpu:~/development/carma-workspace$ docker run -it --rm -v $PWD/c1tenth-driver-wrappers:/home/carma/c1tenth-driver-wrappers quitter.tech/carma-platform:c1tenth-develop  bash
carma@8ebf1060547f:~$ mkdir -p ~/tmp_ws/src
carma@8ebf1060547f:~$ cd ~/tmp_ws/src
carma@8ebf1060547f:~/tmp_ws/src$ ln -s /home/carma/c1tenth-driver-wrappers 
carma@8ebf1060547f:~/tmp_ws/src$ cd ..
carma@8ebf1060547f:~/tmp_ws$ source /opt/carma/install_ros2/setup.bash 
carma@8ebf1060547f:~/tmp_ws$ colcon list
sllidar_ros2_driver_wrapper     src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper (ros.ament_cmake)
carma@8ebf1060547f:~/tmp_ws$ colcon build
Starting >>> sllidar_ros2_driver_wrapper
--- stderr: sllidar_ros2_driver_wrapper                               
In file included from /home/carma/tmp_ws/src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper/include/sllidar_ros2_driver_wrapper/sllidar_ros2_driver_wrapper.hpp:20,
                 from /home/carma/tmp_ws/src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper/src/sllidar_ros2_driver_wrapper.cpp:18:
/home/carma/tmp_ws/src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper/include/sllidar_ros2_driver_wrapper/sllidar_ros2_driver_wrapper_config.hpp:28:5: error: expected ‘{’ before ‘struct’
   28 |     struct Config
      |     ^~~~~~
/home/carma/tmp_ws/src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper/include/sllidar_ros2_driver_wrapper/sllidar_ros2_driver_wrapper_config.hpp:41:1: error: expected declaration before ‘}’ token
   41 | }//namespace sllidar_ros2_driver_wrapper
      | ^
/home/carma/tmp_ws/src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper/src/sllidar_ros2_driver_wrapper.cpp: In member function ‘void sllidar_ros2_driver_wrapper::Node::point_cloud_cb(sensor_msgs::msg::PointCloud2_<std::allocator<void> >::UniquePtr)’:
/home/carma/tmp_ws/src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper/src/sllidar_ros2_driver_wrapper.cpp:30:78: warning: unused parameter ‘msg’ [-Wunused-parameter]
   30 |     void Node::point_cloud_cb(const sensor_msgs::msg::PointCloud2::UniquePtr msg)
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
/home/carma/tmp_ws/src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper/src/sllidar_ros2_driver_wrapper.cpp: In member function ‘virtual carma_ros2_utils::CallbackReturn sllidar_ros2_driver_wrapper::Node::handle_on_configure(const rclcpp_lifecycle::State&)’:
/home/carma/tmp_ws/src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper/src/sllidar_ros2_driver_wrapper.cpp:36:95: warning: unused parameter ‘prev_state’ [-Wunused-parameter]
   36 |     carma_ros2_utils::CallbackReturn Node::handle_on_configure(const rclcpp_lifecycle::State &prev_state)
      |                                                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
/home/carma/tmp_ws/src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper/src/sllidar_ros2_driver_wrapper.cpp: In member function ‘virtual carma_ros2_utils::CallbackReturn sllidar_ros2_driver_wrapper::Node::handle_on_activate(const rclcpp_lifecycle::State&)’:
/home/carma/tmp_ws/src/c1tenth-driver-wrappers/sllidar_ros2_driver_wrapper/src/sllidar_ros2_driver_wrapper.cpp:53:94: warning: unused parameter ‘prev_state’ [-Wunused-parameter]
   53 |     carma_ros2_utils::CallbackReturn Node::handle_on_activate(const rclcpp_lifecycle::State &prev_state)
      |                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
make[2]: *** [CMakeFiles/sllidar_ros2_driver_wrapper_node.dir/build.make:63: CMakeFiles/sllidar_ros2_driver_wrapper_node.dir/src/sllidar_ros2_driver_wrapper.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:107: CMakeFiles/sllidar_ros2_driver_wrapper_node.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
---
Failed   <<< sllidar_ros2_driver_wrapper [28.3s, exited with code 2]
                                 
Summary: 0 packages finished [29.4s]
  1 package failed: sllidar_ros2_driver_wrapper
  1 package had stderr output: sllidar_ros2_driver_wrapper
carma@8ebf1060547f:~/tmp_ws$ 
```

**Comment by asymingt** [comment]
> Andrew should review this

I like the direction it's taking. I'm only going to do a comprehensive review once you have it compiling in a containerized context. As it currently stands there are syntax errors.

File: sllidar_ros2_driver_wrapper/package.xml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  <name>sllidar_ros2_driver_wrapper</name>
+  <version>1.0.0</version>
+  <description>CARMA lidar driver wrapper package</description>
+  <maintainer email="carma@dot.gov">carma</maintainer>
+  <license>Apache 2.0</license>
+
+  <buildtool_depend>ament_cmake_auto</buildtool_depend>
+  <build_depend>carma_cmake_common</build_depend>
+  <build_depend>rcpputils</build_depend>
+  <build_depend>ament_cmake_auto</build_depend>
```
Comment (asymingt) [inline sllidar_ros2_driver_wrapper/package.xml:12]:
```diff
diff --git a/sllidar_ros2_driver_wrapper/package.xml b/sllidar_ros2_driver_wrapper/package.xml
index 0bc9bec..ef1fec5 100644
--- a/sllidar_ros2_driver_wrapper/package.xml
+++ b/sllidar_ros2_driver_wrapper/package.xml
@@ -14,6 +14,7 @@
   <depend>carma_ros2_utils</depend>
   <depend>sensor_msgs</depend>
   <depend>rclcpp</depend>
+  <depend>rclcpp_lifecycle</depend>
   
   <exec_depend>sllidar_ros2</exec_depend>
   <exec_depend>launch</exec_depend>
```

**Comment by nickm1824** [comment]
carma@24e06d6b562c:~/tmp_ws$ ros2 launch sllidar_ros2_driver_wrapper sllidar_ros2_driver_wrapper_launch.py 
[INFO] [launch]: All log files can be found below /home/carma/.ros/log/2023-02-16-15-31-34-537963-24e06d6b562c-383
[INFO] [launch]: Default logging verbosity is set to INFO
Task exception was never retrieved
future: <Task finished name='Task-2' coro=<LaunchService._process_one_event() done, defined at /opt/ros/foxy/lib/python3.8/site-packages/launch/launch_service.py:226> exception=InvalidLaunchFileError('py')>
Traceback (most recent call last):
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_description_sources/any_launch_file_utilities.py", line 53, in get_launch_description_from_any_launch_file
    return loader(launch_file_path)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_description_sources/python_launch_file_utilities.py", line 62, in get_launch_description_from_python_launch_file
    launch_file_module = load_python_launch_file_as_module(python_launch_file_path)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_description_sources/python_launch_file_utilities.py", line 37, in load_python_launch_file_as_module
    loader.exec_module(mod)
  File "<frozen importlib._bootstrap_external>", line 844, in exec_module
  File "<frozen importlib._bootstrap_external>", line 981, in get_code
  File "<frozen importlib._bootstrap_external>", line 911, in source_to_code
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/carma/tmp_ws/install/sllidar_ros2_driver_wrapper/share/sllidar_ros2_driver_wrapper/launch/sllidar_ros2_driver_wrapper_launch.py", line 41
    return launch.LaunchDescription([
    ^
IndentationError: unexpected indent

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_service.py", line 228, in _process_one_event
    await self.__process_event(next_event)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_service.py", line 248, in __process_event
    visit_all_entities_and_collect_futures(entity, self.__context))
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 45, in visit_all_entities_and_collect_futures
    futures_to_return += visit_all_entities_and_collect_futures(sub_entity, context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 45, in visit_all_entities_and_collect_futures
    futures_to_return += visit_all_entities_and_collect_futures(sub_entity, context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 38, in visit_all_entities_and_collect_futures
    sub_entities = entity.visit(context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/action.py", line 108, in visit
    return self.execute(context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/actions/include_launch_description.py", line 130, in execute
    launch_description = self.__launch_description_source.get_launch_description(context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_description_source.py", line 84, in get_launch_description
    self._get_launch_description(self.__expanded_location)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_description_sources/any_launch_description_source.py", line 53, in _get_launch_description
    return get_launch_description_from_any_launch_file(location)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_description_sources/any_launch_file_utilities.py", line 56, in get_launch_description_from_any_launch_file
    raise InvalidLaunchFileError(extension, likely_errors=exceptions)
launch.invalid_launch_file_error.InvalidLaunchFileError: Caught exception when trying to load file of format [py]: unexpected indent (sllidar_ros2_driver_wrapper_launch.py, line 41)

**Pull request merged**